File: p182.java
/* 
 * Solution to Project Euler problem 182
 * By Nayuki Minase
 * 
 * http://nayuki.eigenstate.org/page/project-euler-solutions
 * https://github.com/nayuki/Project-Euler-solutions
 */


public class p182 {
	
	private static int P = 1009;
	private static int Q = 3643;
	private static int TOTIENT = (P - 1) * (Q - 1);
	private static int N = P * Q;
	
	
	public static void main(String[] args) {
		int[] numUnconcealed = new int[TOTIENT];
		for (int e = 0; e < numUnconcealed.length; e++) {
			if (Library.gcd(e, TOTIENT) == 1)
				numUnconcealed[e] = countUnconcealed(e);
			else
				numUnconcealed[e] = Integer.MAX_VALUE;
		}
		
		int minUnconcealed = Integer.MAX_VALUE;
		for (int x : numUnconcealed)
			minUnconcealed = Math.min(x, minUnconcealed);
		
		long sum = 0;
		for (int i = 0; i < numUnconcealed.length; i++) {
			if (numUnconcealed[i] == minUnconcealed)
				sum += i;
		}
		System.out.println(sum);
	}
	
	
	private static int countUnconcealed(int e) {
		int count0 = 0;
		for (int m = 0; m < P; m++) {
			if (powMod(m, e, P) == m)
				count0++;
		}
		int count1 = 0;
		for (int m = 0; m < Q; m++) {
			if (powMod(m, e, Q) == m)
				count1++;
		}
		return count0 * count1;
	}
	
	
	private static int countUnconcealedSlow(int e) {
		int count = 0;
		for (int m = 0; m < N; m++) {
			if (powMod(m, e, N) == m)
				count++;
		}
		return count;
	}
	
	
	private static int powMod(int x, int y, int m) {
		if (y < 0)
			throw new IllegalArgumentException();
		int z = 1;
		for (; y != 0; y >>>= 1, x = (int)((long)x * x % m)) {
			if ((y & 1) != 0)
				z = (int)((long)z * x % m);
		}
		return z;
	}
	
}

File: haskell/p182.hs
{- 
 - Solution to Project Euler problem 182
 - Copyright (c) Project Nayuki. All rights reserved.
 - 
 - https://www.nayuki.io/page/project-euler-solutions
 - https://github.com/nayuki/Project-Euler-solutions
 -}

import qualified EulerLib


p = 1009 :: Integer
q = 3643 :: Integer
main = putStrLn (show ans)
numUnconcealedP = countAllUnconcealed p
numUnconcealedQ = countAllUnconcealed q
minUnconcealedP = minimum numUnconcealedP
minUnconcealedQ = minimum numUnconcealedQ
totient = (p - 1) * (q - 1)
ans = let
		cond e = (numUnconcealedP !! (fromInteger (mod e (p - 1)))) == minUnconcealedP
		      && (numUnconcealedQ !! (fromInteger (mod e (q - 1)))) == minUnconcealedQ
	in sum (filter cond [0 .. (totient - 1)])


countAllUnconcealed :: Integer -> [Integer]
countAllUnconcealed prime = let
		func e = if (gcd e (prime - 1)) == 1
			then (countUnconcealed prime e)
			else 10^20  -- Sentinel
	in map func [0 .. (prime - 2)]


countUnconcealed :: Integer -> Integer -> Integer
countUnconcealed modulus e = fromIntegral (EulerLib.count (\m -> (powMod m e modulus) == m) [0 .. (modulus - 1)])


powMod :: Integer -> Integer -> Integer -> Integer
powMod _ 0 m = mod 1 m
powMod x y m
	| mod y 2 == 0 = temp
	| otherwise    = mod (temp * x) m
	where
		temp = powMod (mod (x * x) m) (div y 2) m
